coral-presto/src/main/java/com/linkedin/coral/presto/rel2presto/functions/PrestoCardinalityFunction.java

```diff
@@ -0,0 +1,40 @@
+/**
+ * Copyright 2018-2021 LinkedIn Corporation. All rights reserved.
+ * Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+package com.linkedin.coral.presto.rel2presto.functions;
+
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.ReturnTypes;
+
+
+/**
+ * Represents the CARDINALITY function in Presto
+ *
+ * Identical the CARDINALITY SQL function, but CARDINALITY in PRESTO returns a BIGINT so we
+ * need to cast it to an INTEGER when unparsing
+ */
+public class PrestoCardinalityFunction extends SqlFunction {
+  public static final PrestoCardinalityFunction INSTANCE = new PrestoCardinalityFunction();
+
+  public PrestoCardinalityFunction() {
+    super("CARDINALITY", SqlKind.OTHER_FUNCTION, ReturnTypes.INTEGER_NULLABLE, null, OperandTypes.COLLECTION_OR_MAP,
+        SqlFunctionCategory.SYSTEM);
+  }
+
+  @Override
+  public void unparse(SqlWriter writer, SqlCall call, int leftPrec, int rightPrec) {
+    final SqlWriter.Frame frame = writer.startFunCall(SqlStdOperatorTable.CAST.getName());
+    super.unparse(writer, call, leftPrec, rightPrec);
+    writer.sep("AS");
+    writer.sep("INTEGER");
+    writer.endFunCall(frame);
+  }
```

Comment:
Does it make sense to add the `CAST` in `RelToPrestoConverter`?

Reply:
That was my first option as well, however the issue is that you have `x` with is a cardinality call, the result will be `CAST(x AS INT)` and now the converter will try to visit the `CAST` and the run into `x` again and try to add a cast again going into infinite recursion. Unfortunately when the call is delegated to our visitor, we don't have any context about the parent to be able to break the recursion.

Reply:
Another option is to add it to the `CalcitePrestoUDFMap`, and apply an operand transformer? For example, the `Truncate` function goes through a transformation:
```
    createUDFMapEntry(UDF_MAP, SqlStdOperatorTable.TRUNCATE, 2, "TRUNCATE",
        "[{\"op\":\"*\",\"operands\":[{\"input\":1},{\"op\":\"^\",\"operands\":[{\"value\":10},{\"input\":2}]}]}]",
        "{\"op\":\"/\",\"operands\":[{\"input\":0},{\"op\":\"^\",\"operands\":[{\"value\":10},{\"input\":2}]}]}");
```
Operand Transformer is just a convenience method to return a new `RexNode`. If the string based operand transformation of the operand transformer is not expressive enough for this case, it may be fair to use a custom `RexNode`.

Reply:
I had also tried this by creating a number operator called CAST as a result transformer. It has the same problem of infinite recursion.

Reply:
Discussed another approach offline: infinite recursion is due to the collision between the Calcite name of the operator and the Presto name (`Truncate` in this case).  We have another symptom of this problem in https://github.com/linkedin/coral/pull/23. I think Coral needs its unambiguous operator names to avoid cases like this and https://github.com/linkedin/coral/pull/23. The same suggestion in [this comment](https://github.com/linkedin/coral/pull/23#issuecomment-774927641) may help here.